### PR TITLE
[12.0] [ADD] spp_portal_customer_wallet

### PR DESCRIPTION
**Pull request by carmenbianca**
## Odoo task (if applicable)

https://gestion.coopiteasy.be/web#id=10213&action=479&model=project.task&view_type=form&menu_id=536

## Checklist before approval

- [x] Tests are present (or not needed).
- [x] Credits/copyright have been changed correctly.
- [ ] Change log snippet is present.
- [ ] (If a new module) Moving this to OCA has been considered.

### CODE REVIEW

File: spp_portal_customer_wallet/controllers/main.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        12: _("December"),
+    }
+    return MONTHS[number]
+
+
+def sums_of_years(per_month):
+    result = defaultdict(int)
+    for key, value in per_month.items():
+        result[key[0]] += value
+    return dict(result)
```
Comment (victor-champonnois) [inline spp_portal_customer_wallet/controllers/main.py:35]: Sorry I don't understand what this function does. Is there a way to make it more explicit ? (a comment or clearer names ?)

File: spp_portal_customer_wallet/controllers/main.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+
+class PortalPosOrderAmount(CustomerPortal):
+    def _prepare_portal_layout_values(self):
+        values = super(PortalPosOrderAmount, self)._prepare_portal_layout_values()
```
Comment (victor-champonnois) [inline spp_portal_customer_wallet/controllers/main.py:40]: Some say it's best to just use `super()` in that case (without parameter). It would be maybe more consistent with our codebase. I don't really care though :)

File: spp_portal_customer_wallet/controllers/main.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+                )
+                years.add(year)
+            ordered.append(
+                {
+                    "month": f"{translate_month(month)} {year}",
+                    "amount": value,
+                    "foodprint_amount": foodprint_per_month.get(key, 0),
+                }
+            )
+
```
Comment (victor-champonnois) [inline spp_portal_customer_wallet/controllers/main.py:70]: It's very hard to understand this code (at least for me). I think it should be encapsulated in a method (or methods) which has a name describing what it does, otherwise it's just too much for my brain :/

Comment (victor-champonnois) [reply spp_portal_customer_wallet/controllers/main.py:70]: Or you can comments if it's too much refactoring, I just need to be explained in plain english what this does. I think the "per_month" name is very obscure, maybe that is the culprit.

File: spp_portal_customer_wallet/models/res_partner.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+from odoo import models
+
+
+class ResPartner(models.Model):
+    _inherit = "res.partner"
+
+    def customer_wallet_payments_per_month(self):
+        """Return a dictionary with (year, month) keys. The value is the amount
+        spent using the customer wallet for every month.
+        """
```
Comment (victor-champonnois) [inline spp_portal_customer_wallet/models/res_partner.py:16]: OK that makes it a bit clearer :)

**Review by victor-champonnois — CHANGES_REQUESTED** [review]
Very complex so I was lost on the logic. I will trust the functionnal tests on that. However I think the code should be made easier to read by commenting or encapsulating in smaller methods.

**Comment by victor-champonnois** [comment]
@carmenbianca I also think that this module could have a more generic name no ? It's really a new functionality that could be reused.

**Comment by huguesdk** [comment]
@victor-champonnois thanks for the review. i’m afraid we’re too close to the release for changes. i skimmed through the code and didn’t spot anything dangerous or weird. since it’s a new module and it’s custom (so will be used by only one client), i think it’s ok to release it as is, and your asked changes should be done later in another branch.

**Comment by victor-champonnois** [comment]
@huguesdk You are right as always hugues :) I approve the PR and get out of the way ;)

**Review by victor-champonnois — APPROVED** [review]

**Comment by huguesdk** [comment]
sorry if i wasn’t clear. i didn’t mean to approve it, as i agree with your remarks. i just don’t think they should block the release, only that they should be handled later.

**Comment by victor-champonnois** [comment]
Ah ok I didn't think of this option. Let's do it like that then :)
